###
### Batocera.PLUS
### Alexandre Freire dos Santos
###
### This patch fixes problems with 2D graphics of several games. 
### It is forbidden to use this patch without giving due credit to the Batocera.PLUS project. 
### 
diff --git a/core/rend/gles/gles.cpp b/core/rend/gles/gles.cpp
index 3961f54..164ff86 100644
--- a/core/rend/gles/gles.cpp
+++ b/core/rend/gles/gles.cpp
@@ -194,8 +194,7 @@ highp vec4 fog_clamp(highp vec4 col)
 
 lowp vec4 palettePixel(highp vec2 coords)
 {
-   highp int color_idx = int(floor(texture(tex, coords).FOG_CHANNEL * 255.0 + 0.5)) + palette_index;
-   highp vec2 c = vec2(mod(float(color_idx), 32.0) / 31.0, float(color_idx / 32) / 31.0);
+   highp vec2 c = vec2((texture(tex, coords).FOG_CHANNEL * 255.0 + float(palette_index)) / 1023.0, 0.5);
 	return texture(palette, c);
 }
 
@@ -754,7 +753,7 @@ void UpdatePaletteTexture(GLenum texture_slot)
 		glcache.BindTexture(GL_TEXTURE_2D, paletteTextureId);
 
    glPixelStorei(GL_UNPACK_ALIGNMENT, 1);
-	glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, 32, 32, 0, GL_RGBA, GL_UNSIGNED_BYTE, palette32_ram);
+	glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, 1024, 1, 0, GL_RGBA, GL_UNSIGNED_BYTE, palette32_ram);
 	glCheck();
 
 	glActiveTexture(GL_TEXTURE0);
